# **Encoding and Feature Engineering:**
* before starting with encoding we will be deleting the last row because we have total rows in odd number 1273 for model training keeping even rows will be *better*
* Creating new data frame as encoded data fraem by from copy of new_df

In [ ]:
# Loading cleaned and Preprocessed Data set saved from Part-1 notebook

from google.colab import files
uploaded = files.upload()

Saving new_laptop.xlsx to new_laptop.xlsx


In [ ]:
import pandas as pd
import numpy as np
import sklearn as sk
import re

In [ ]:
# loading the dataset
new_df = pd.read_excel("new_laptop.xlsx")

In [ ]:
# making copy of the data
copy_new_df = new_df.copy()

In [ ]:
new_df.head()

Company   TypeName  Inches        Price  Weight_kg IPS_Panel Touchscreen  \
0   Apple  Ultrabook    13.3   71378.6832       1.37       Yes          No   
1   Apple  Ultrabook    13.3   47895.5232       1.34        No          No   
2      HP   Notebook    15.6   30636.0000       1.86        No          No   
3   Apple  Ultrabook    15.4  135195.3360       1.83       Yes          No   
4   Apple  Ultrabook    13.3   96095.8080       1.37       Yes          No   

  Resolution  PPI    Display_Type  ... RAM_GB Memory_1_Size_GB  Memory_1_Type  \
0  2560x1600  227  Retina Display  ...      8              128            SSD   
1   1440x900  128             HD+  ...      8              128  Flash Storage   
2  1920x1080  141         Full HD  ...      8              256            SSD   
3  2880x1800  221  Retina Display  ...     16              512            SSD   
4  2560x1600  227  Retina Display  ...      8              256            SSD   

   Memory_2_Size_GB  Memory_2_Type GPU_Brand  GPU_Model   GPU_Model_series  \
0                 0  Not Specified     Intel       Iris  Plus Graphics 640   
1                 0  Not Specified     Intel         HD      Graphics 6000   
2                 0  Not Specified     Intel         HD       Graphics 620   
3                 0  Not Specified       AMD     Radeon            Pro 455   
4                 0  Not Specified     Intel       Iris  Plus Graphics 650   

  OS_Brand     OS_Version  
0    macOS  Not Specified  
1    macOS  Not Specified  
2    No OS  Not Specified  
3    macOS  Not Specified  
4    macOS  Not Specified  

[5 rows x 23 columns]

In [ ]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1273 entries, 0 to 1272
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1273 non-null   object 
 1   TypeName          1273 non-null   object 
 2   Inches            1273 non-null   float64
 3   Price             1273 non-null   float64
 4   Weight_kg         1273 non-null   float64
 5   IPS_Panel         1273 non-null   object 
 6   Touchscreen       1273 non-null   object 
 7   Resolution        1273 non-null   object 
 8   PPI               1273 non-null   int64  
 9   Display_Type      1273 non-null   object 
 10  CPU_Brand         1273 non-null   object 
 11  CPU_Series        1273 non-null   object 
 12  CPU_Speed_GHz     1273 non-null   float64
 13  RAM_GB            1273 non-null   int64  
 14  Memory_1_Size_GB  1273 non-null   int64  
 15  Memory_1_Type     1273 non-null   object 
 16  Memory_2_Size_GB  1273 non-null   int64  


# Feature Selection

In [ ]:
# Removing Last row for better model training as our dataset with total odd no. of columns
new_df = new_df.drop(new_df.index[-1])

In [ ]:
# Creating a new column as Processor by concatenating CPU_Brand and CPU_Series and if CPU series column contains not specified data, then returning only CPU brand
def concat_cpu_brand_series(brand, series):
    if series == "Not Specified":
        return brand
    else:
        return brand + ' ' + series
new_df['Processor'] = new_df.apply(lambda row: concat_cpu_brand_series(row['CPU_Brand'], row['CPU_Series']), axis=1)

In [ ]:
# df['Total_Memory_Size_GB'] = df['Memory_1_Size_GB'] + df['Memory_2_Size_GB']

In [ ]:
# Creating a new column as Graphic_Card by concatenating GPU_Brand and GPU_Model and if GPU moel column contains not specified data, then returning only GPU brand
def concat_gpu_brand_model(brand, model):
    if model == "Not Specified":
        return brand
    else:
        return brand + ' ' + model
new_df['Graphic_Card'] = new_df.apply(lambda row: concat_gpu_brand_model(row['GPU_Brand'], row['GPU_Model']), axis=1)

In [ ]:
# Dropping those column and also extra feature column which are not required for price prediction model
new_df.drop(columns=['Resolution', 'CPU_Brand', 'CPU_Series', 'GPU_Brand', 'GPU_Model'], inplace=True)

In [ ]:
new_df.sample(5)

Company   TypeName  Inches       Price  Weight_kg IPS_Panel Touchscreen  \
45     Asus     Gaming    17.3   69210.720       3.20        No          No   
497  Lenovo     Gaming    15.6   61218.720       2.40       Yes          No   
830    Dell   Notebook    17.3   46087.200       2.83        No          No   
864    Dell   Notebook    15.6   24279.696       2.18        No          No   
477    Asus  Ultrabook    12.5  103896.000       0.91        No          No   

     PPI Display_Type  CPU_Speed_GHz  RAM_GB  Memory_1_Size_GB Memory_1_Type  \
45   127      Full HD            3.0       8               256           SSD   
497  141      Full HD            2.8       8               256           SSD   
830  127      Full HD            2.7       8              1024           HDD   
864  100           HD            2.0       4              1024           HDD   
477  176      Full HD            2.7      16               512           SSD   

     Memory_2_Size_GB  Memory_2_Type GPU_Model_series OS_Brand     OS_Version  \
45               1024            HDD           RX 580  Windows             10   
497              1024            HDD         GTX 1060    No OS  Not Specified   
830                 0  Not Specified          R7 M445    Linux  Not Specified   
864                 0  Not Specified         R5 M420X    Linux  Not Specified   
477                 0  Not Specified     Graphics 620  Windows             10   

         Processor    Graphic_Card  
45       AMD Ryzen      AMD Radeon  
497  Intel Core i7  Nvidia GeForce  
830  Intel Core i7      AMD Radeon  
864  Intel Core i3      AMD Radeon  
477  Intel Core i7        Intel HD

# Creating a new Dataframe after feature selection from new df making copy of it as encoded df

In [ ]:
# Creating new df as encoded df
encoded_df = new_df.copy()

In [ ]:
encoded_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1272 entries, 0 to 1271
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1272 non-null   object 
 1   TypeName          1272 non-null   object 
 2   Inches            1272 non-null   float64
 3   Price             1272 non-null   float64
 4   Weight_kg         1272 non-null   float64
 5   IPS_Panel         1272 non-null   object 
 6   Touchscreen       1272 non-null   object 
 7   PPI               1272 non-null   int64  
 8   Display_Type      1272 non-null   object 
 9   CPU_Speed_GHz     1272 non-null   float64
 10  RAM_GB            1272 non-null   int64  
 11  Memory_1_Size_GB  1272 non-null   int64  
 12  Memory_1_Type     1272 non-null   object 
 13  Memory_2_Size_GB  1272 non-null   int64  
 14  Memory_2_Type     1272 non-null   object 
 15  GPU_Model_series  1272 non-null   object 
 16  OS_Brand          1272 non-null   object 


**Total Columns : 20**

**Total rows : 1272**

**Total Numerical column : 8**
['Inches', 'Price', 'Weight_kg', 'PPI', 'CPU_Speed_GHz', 'RAM_GB', 'Memory_1_Size_GB', 'Memory_2_Size_GB']

**Total Categorial column : 12**
[Company, TypeName, IPS_Panel, Touchscreen, Display_Type, Processor, Memory_1_Type, Memory_2_Type, Graphic_Card, GPU_Performance, OS_Brand, OS_Version]
       
          
    
       
              
         
         
            
          

In [ ]:
# Checking unique values in all categorial column for better understading before doing encoding
numerical_columns = ['Inches', 'Price', 'Weight_kg', 'PPI', 'CPU_Speed_GHz', 'RAM_GB', 'Memory_1_Size_GB', 'Memory_2_Size_GB']
for column in encoded_df.columns:
    if column in numerical_columns:
        continue  # Skipping numerical columns
    unique_values = encoded_df[column].unique()
    total_variants = len(unique_values)
    print(f"Column '{column}':")
    print(f"Unique values: {unique_values}")
    print(f"Total variants: {total_variants}")
    print("========================================")

Column 'Company':
Unique values: ['Apple' 'HP' 'Acer' 'Asus' 'Dell' 'Lenovo' 'Chuwi' 'MSI' 'Microsoft'
 'Toshiba' 'Huawei' 'Xiaomi' 'Vero' 'Razer' 'Mediacom' 'Samsung' 'Google'
 'Fujitsu' 'LG']
Total variants: 19
Column 'TypeName':
Unique values: ['Ultrabook' 'Notebook' 'Gaming' '2 in 1 Convertible' 'Workstation'
 'Netbook']
Total variants: 6
Column 'IPS_Panel':
Unique values: ['Yes' 'No']
Total variants: 2
Column 'Touchscreen':
Unique values: ['No' 'Yes']
Total variants: 2
Column 'Display_Type':
Unique values: ['Retina Display' 'HD+' 'Full HD' 'HD' 'Quad HD' 'Not Specified'
 '4K Ultra HD']
Total variants: 7
Column 'Memory_1_Type':
Unique values: ['SSD' 'Flash Storage' 'HDD' 'Hybrid']
Total variants: 4
Column 'Memory_2_Type':
Unique values: ['Not Specified' 'HDD' 'SSD' 'Hybrid']
Total variants: 4
Column 'GPU_Model_series':
Unique values: ['Plus Graphics 640' 'Graphics 6000' 'Graphics 620' 'Pro 455'
 'Plus Graphics 650' 'R5' 'Pro Graphics' 'MX150' 'Graphics 520' 'Pro 555'
 'R5 M430' 'Gr

## Defining Encoding technique to the each categorial columns
**Nominal Column :** Company, TypeName, IPS_Panel, Touchscreen, Memory_1_Type, Memory_2_Type, OS_Brand,

**Ordinal Column :** Display_Type, OS_Version, Graphic_Card, Processor, GPU_Model_series

* **Label encoding:** Company, TypeName, IPS_Panel, Touchscreen, OS_Brand, Graphic_Card, OS_Version, Memory_1_Type, Memory_2_Type, Processor, GPU_Model_series

* **Ordinal encoding:** Display_Type


In [ ]:
# data before encoding
encoded_df.head()

Company   TypeName  Inches        Price  Weight_kg IPS_Panel Touchscreen  \
0   Apple  Ultrabook    13.3   71378.6832       1.37       Yes          No   
1   Apple  Ultrabook    13.3   47895.5232       1.34        No          No   
2      HP   Notebook    15.6   30636.0000       1.86        No          No   
3   Apple  Ultrabook    15.4  135195.3360       1.83       Yes          No   
4   Apple  Ultrabook    13.3   96095.8080       1.37       Yes          No   

   PPI    Display_Type  CPU_Speed_GHz  RAM_GB  Memory_1_Size_GB  \
0  227  Retina Display            2.3       8               128   
1  128             HD+            1.8       8               128   
2  141         Full HD            2.5       8               256   
3  221  Retina Display            2.7      16               512   
4  227  Retina Display            3.1       8               256   

   Memory_1_Type  Memory_2_Size_GB  Memory_2_Type   GPU_Model_series OS_Brand  \
0            SSD                 0  Not Specified  Plus Graphics 640    macOS   
1  Flash Storage                 0  Not Specified      Graphics 6000    macOS   
2            SSD                 0  Not Specified       Graphics 620    No OS   
3            SSD                 0  Not Specified            Pro 455    macOS   
4            SSD                 0  Not Specified  Plus Graphics 650    macOS   

      OS_Version      Processor Graphic_Card  
0  Not Specified  Intel Core i5   Intel Iris  
1  Not Specified  Intel Core i5     Intel HD  
2  Not Specified  Intel Core i5     Intel HD  
3  Not Specified  Intel Core i7   AMD Radeon  
4  Not Specified  Intel Core i5   Intel Iris

In [ ]:
# Label encoding
from sklearn.preprocessing import LabelEncoder
# column to encode
label_encode_columns = ['Company', 'TypeName', 'IPS_Panel', 'Touchscreen', 'Memory_1_Type', 'Memory_2_Type', 'OS_Brand', 'OS_Version', 'Processor', 'Graphic_Card', 'GPU_Model_series', 'Display_Type']

label_encoders = {}
for column in label_encode_columns:
    label_encoders[column] = LabelEncoder()
    encoded_df[column] = label_encoders[column].fit_transform(encoded_df[column])

In [ ]:
# Data after encoding
encoded_df.head()

Company  TypeName  Inches        Price  Weight_kg  IPS_Panel  Touchscreen  \
0        1         4    13.3   71378.6832       1.37          1            0   
1        1         4    13.3   47895.5232       1.34          0            0   
2        7         3    15.6   30636.0000       1.86          0            0   
3        1         4    15.4  135195.3360       1.83          1            0   
4        1         4    13.3   96095.8080       1.37          1            0   

   PPI  Display_Type  CPU_Speed_GHz  RAM_GB  Memory_1_Size_GB  Memory_1_Type  \
0  227             6            2.3       8               128              3   
1  128             3            1.8       8               128              0   
2  141             1            2.5       8               256              3   
3  221             6            2.7      16               512              3   
4  227             6            3.1       8               256              3   

   Memory_2_Size_GB  Memory_2_Type  GPU_Model_series  OS_Brand  OS_Version  \
0                 0              2                65         5           3   
1                 0              2                49         5           3   
2                 0              2                51         3           3   
3                 0              2                67         5           3   
4                 0              2                66         5           3   

   Processor  Graphic_Card  
0         14             6  
1         14             5  
2         14             5  
3         15             2  
4         14             6

In [ ]:
encoded_df.corr()['Price']

Company             0.139671
TypeName           -0.125068
Inches              0.064929
Price               1.000000
Weight_kg           0.151063
IPS_Panel           0.254475
Touchscreen         0.188743
PPI                 0.476395
Display_Type       -0.050561
CPU_Speed_GHz       0.428315
RAM_GB              0.685454
Memory_1_Size_GB   -0.128967
Memory_1_Type       0.508922
Memory_2_Size_GB    0.296554
Memory_2_Type      -0.302613
GPU_Model_series   -0.101600
OS_Brand            0.264136
OS_Version         -0.165992
Processor           0.414500
Graphic_Card        0.379119
Name: Price, dtype: float64

In [ ]:
# Save the DataFrame to an Excel file
encoded_df.to_excel('encoded_laptop.xlsx', index=False)

In [ ]:
encoded_df['OS_Version'].unique()

array([3, 0, 4, 1, 2])

# **Model Development :**

In [ ]:
# Creating a copy of encoded data
model_df = encoded_df.copy()

In [ ]:
model_df

Company  TypeName  Inches        Price  Weight_kg  IPS_Panel  \
0           1         4    13.3   71378.6832       1.37          1   
1           1         4    13.3   47895.5232       1.34          0   
2           7         3    15.6   30636.0000       1.86          0   
3           1         4    15.4  135195.3360       1.83          1   
4           1         4    13.3   96095.8080       1.37          1   
...       ...       ...     ...          ...        ...        ...   
1267        2         3    15.6   38378.6496       2.20          0   
1268       10         0    14.0   33992.6400       1.80          1   
1269       10         0    13.3   79866.7200       1.30          1   
1270       10         3    14.0   12201.1200       1.50          0   
1271        7         3    15.6   40705.9200       2.19          0   

      Touchscreen  PPI  Display_Type  CPU_Speed_GHz  RAM_GB  Memory_1_Size_GB  \
0               0  227             6            2.3       8               128   
1               0  128             3            1.8       8               128   
2               0  141             1            2.5       8               256   
3               0  221             6            2.7      16               512   
4               0  227             6            3.1       8               256   
...           ...  ...           ...            ...     ...               ...   
1267            0  100             2            2.5       4               500   
1268            1  157             1            2.5       4               128   
1269            1  276             5            2.5      16               512   
1270            0  112             2            1.6       2                64   
1271            0  100             2            2.5       6              1024   

      Memory_1_Type  Memory_2_Size_GB  Memory_2_Type  GPU_Model_series  \
0                 3                 0              2                65   
1                 0                 0              2                49   
2                 3                 0              2                51   
3                 3                 0              2                67   
4                 3                 0              2                66   
...             ...               ...            ...               ...   
1267              1                 0              2                 6   
1268              3                 0              2                44   
1269              3                 0              2                44   
1270              0                 0              2                37   
1271              1                 0              2                81   

      OS_Brand  OS_Version  Processor  Graphic_Card  
0            5           3         14             6  
1            5           3         14             5  
2            3           3         14             5  
3            5           3         15             2  
4            5           3         14             6  
...        ...         ...        ...           ...  
1267         4           0         15             9  
1268         4           0         15             5  
1269         4           0         15             5  
1270         4           0         10             5  
1271         4           0         15             2  

[1272 rows x 20 columns]

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
# Splitting features and target variable
X = model_df.drop('Price', axis=1)     # For feature variable
Y = model_df['Price']               # Target variable

In [ ]:
# Splitting the dataset into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=40)

In [ ]:
X_train

Company  TypeName  Inches  Weight_kg  IPS_Panel  Touchscreen  PPI  \
1222       10         3    15.6       2.40          0            0  141   
452        10         3    14.0       1.50          1            0  157   
888         4         3    14.0       1.60          0            0  157   
304         0         3    15.6       2.23          0            0  141   
816        10         3    14.0       1.50          0            0  112   
...       ...       ...     ...        ...        ...          ...  ...   
626        10         4    14.0       1.14          1            0  157   
1016       10         3    15.6       2.30          0            0  141   
165         7         3    17.3       2.71          1            0  127   
7           1         4    13.3       1.34          0            0  128   
219         2         3    15.6       2.00          0            0  141   

      Display_Type  CPU_Speed_GHz  RAM_GB  Memory_1_Size_GB  Memory_1_Type  \
1222             1            2.3       6               128              3   
452              1            2.7       8               512              3   
888              1            2.8       8               256              3   
304              1            2.2       8               256              3   
816              2            1.6       4                16              0   
...            ...            ...     ...               ...            ...   
626              1            2.7      16               512              3   
1016             1            2.3       8               256              3   
165              1            2.5       6              2048              1   
7                3            1.8       8               256              0   
219              1            2.5       8               256              3   

      Memory_2_Size_GB  Memory_2_Type  GPU_Model_series  OS_Brand  OS_Version  \
1222                 0              2                 7         4           0   
452                  0              2                91         3           3   
888                  0              2                51         4           0   
304                  0              2                77         4           0   
816                  0              2                38         1           3   
...                ...            ...               ...       ...         ...   
626                  0              2                51         4           0   
1016                 0              2                44         4           0   
165                  0              2                 2         4           0   
7                    0              2                49         5           3   
219                  0              2                 7         4           0   

      Processor  Graphic_Card  
1222         13             9  
452          15             2  
888          14             5  
304           5             2  
816          11             5  
...         ...           ...  
626          15             5  
1016         14             5  
165          14             2  
7            14             5  
219          14             9  

[1017 rows x 19 columns]

In [ ]:
X_test

Company  TypeName  Inches  Weight_kg  IPS_Panel  Touchscreen  PPI  \
487       10         3    15.6       2.30          1            0  141   
970        0         3    15.6       2.40          0            0  100   
410        3         3    12.3       1.40          1            0  267   
721        4         3    15.6       2.00          0            0  141   
482        2         1    17.3       3.55          1            1  127   
..       ...       ...     ...        ...        ...          ...  ...   
563       16         4    14.0       1.24          0            1  157   
128        0         3    15.6       3.55          0            0  141   
963        4         3    14.0       1.60          0            0  157   
71         0         3    14.0       2.10          0            0  112   
929       10         4    14.0       1.70          1            0  157   

     Display_Type  CPU_Speed_GHz  RAM_GB  Memory_1_Size_GB  Memory_1_Type  \
487             1            2.5       8               256              3   
970             2            1.1       4               500              1   
410             6            1.1       6                64              0   
721             1            2.7       8               128              3   
482             1            2.6      24               256              3   
..            ...            ...     ...               ...            ...   
563             1            2.7      16               512              3   
128             1            2.7       4               256              3   
963             1            2.5       8               128              3   
71              2            2.0       8              1024              1   
929             1            2.6       8               256              3   

     Memory_2_Size_GB  Memory_2_Type  GPU_Model_series  OS_Brand  OS_Version  \
487                 0              2                51         4           0   
970                 0              2                41         4           0   
410                 0              2                40         4           0   
721              1024              0                11         4           0   
482              1024              0                29         4           0   
..                ...            ...               ...       ...         ...   
563                 0              2                51         4           0   
128                 0              2                63         4           0   
963                 0              2                51         4           0   
71                  0              2                44         4           0   
929                 0              2                44         4           0   

     Processor  Graphic_Card  
487         14             5  
970         17             5  
410         11             5  
721         15             9  
482         15             9  
..         ...           ...  
563         15             5  
128         13             9  
963         14             5  
71          13             5  
929         15             5  

[255 rows x 19 columns]

In [ ]:
Y_train

1222     30316.320
452      42570.720
888      64202.400
304      24988.320
816      17316.000
           ...    
626     130536.000
1016     55922.688
165      37242.720
7        61735.536
219      40972.320
Name: Price, Length: 1017, dtype: float64

In [ ]:
Y_test

487    48538.0800
970    20246.4000
410    23922.7200
721    64308.9600
482    67612.3200
          ...    
563    99367.2000
128    30476.1600
963    59620.3200
71     20725.9200
929    79333.3872
Name: Price, Length: 255, dtype: float64

In [ ]:
# taking this Model for training and evaluation
from sklearn.linear_model import LinearRegression, ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor

In [ ]:
# Initializing models
linear = LinearRegression()
elastic_net = ElasticNet()
decision_tree = DecisionTreeRegressor()
random_forest = RandomForestRegressor()
gradient_boosting = GradientBoostingRegressor()
svm = SVR()
knn = KNeighborsRegressor()
xgb = XGBRegressor()

In [ ]:
# Training the models
linear.fit(X_train, Y_train)
elastic_net.fit(X_train, Y_train)
decision_tree.fit(X_train, Y_train)
random_forest.fit(X_train, Y_train)
gradient_boosting.fit(X_train, Y_train)
svm.fit(X_train, Y_train)
knn.fit(X_train, Y_train)
xgb.fit(X_train, Y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=None, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

In [ ]:
# Making predictions from all models as Y-pred
linear_Y_pred = linear.predict(X_test)
elastic_net_Y_pred = elastic_net.predict(X_test)
decision_tree_Y_pred = decision_tree.predict(X_test)
random_forest_Y_pred = random_forest.predict(X_test)
gradient_boosting_Y_pred = gradient_boosting.predict(X_test)
svm_Y_pred = svm.predict(X_test)
knn_Y_pred = knn.predict(X_test)
xgb_Y_pred = xgb.predict(X_test)

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [ ]:
# Evaluation of the models
def evaluate_model(name, Y_true, Y_pred):
    mae = mean_absolute_error(Y_true, Y_pred)
    mse = mean_squared_error(Y_true, Y_pred)
    r2 = r2_score(Y_true, Y_pred)
    print(f'{name} Model:')
    print(f'Mean Absolute Error: {mae:.2f}')
    print(f'Mean Squared Error: {mse:.2f}')
    print(f'R-squared Score: {r2:.2f}')
    print("-" * 60)

evaluate_model('Linear Regression', Y_test, linear_Y_pred)
evaluate_model('Elastic Net', Y_test, elastic_net_Y_pred)
evaluate_model('Decision Tree Regression', Y_test, decision_tree_Y_pred)
evaluate_model('Random Forest Regression', Y_test, random_forest_Y_pred)
evaluate_model('Gradient Boosting Regression', Y_test, gradient_boosting_Y_pred)
evaluate_model('Support Vector Machine (SVM) Regression', Y_test, svm_Y_pred)
evaluate_model('K-Nearest Neighbors (KNN) Regression', Y_test, knn_Y_pred)
evaluate_model('XGBoost Regression', Y_test, xgb_Y_pred)

Linear Regression Model:
Mean Absolute Error: 16090.85
Mean Squared Error: 504694909.00
R-squared Score: 0.65
------------------------------------------------------------
Elastic Net Model:
Mean Absolute Error: 15833.96
Mean Squared Error: 513315430.59
R-squared Score: 0.64
------------------------------------------------------------
Decision Tree Regression Model:
Mean Absolute Error: 13848.11
Mean Squared Error: 507062309.96
R-squared Score: 0.65
------------------------------------------------------------
Random Forest Regression Model:
Mean Absolute Error: 10548.78
Mean Squared Error: 296586778.94
R-squared Score: 0.79
------------------------------------------------------------
Gradient Boosting Regression Model:
Mean Absolute Error: 10753.48
Mean Squared Error: 271609828.69
R-squared Score: 0.81
------------------------------------------------------------
Support Vector Machine (SVM) Regression Model:
Mean Absolute Error: 28345.24
Mean Squared Error: 1556992952.09
R-squared Score

From above evaluation of the models according to Mean Absolute Error**(MAE)**, Mean Squared Error**(MSE)** and R-Squared Score**(R^2 score**), we come to **Conclusion** that **XGBoost Regression Model** performs well compare to other models

# **Hyperparameter Tuning:**

In [ ]:
# from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from xgboost import XGBRegressor

In [ ]:
# Defining the hyperparameters grid
param_grid = {
    'n_estimators': [200, 240, 250], # Number of trees in XGBoost
    'learning_rate': [0.1, 0.15, 0.2], # Learning rate
    'max_depth': [4, 4.5, 5], # Maximum depth of each tree
    'min_child_weight': [3.5, 4, 4.5], # Minimum sum of instance weight needed in a child
    'gamma': [0.3, 0.35, 0.4], # Minimum loss reduction required to make a further partition on a leaf node
    'subsample': [0.9, 0.95, 1.0], # Subsample ratio of the training instances
    'colsample_bytree': [1.0, 1.0, 1.5], # Subsample ratio of columns when constructing each tree
}

# Initializing the XGBoostRegressor
xgb = XGBRegressor()

# Perform Grid Search with cross-validation
random_search = RandomizedSearchCV(estimator=xgb, param_distributions=param_grid, n_iter=100, cv=5, scoring='neg_mean_squared_error', random_state=42)
# Performing grid search
random_result = random_search.fit(X_train, Y_train)

# Get the best hyperparameters and the best score
best_params = random_result.best_params_
print("Best Hyperparameters:", best_params)

best_score = random_result.best_score_
print("Best Score (neg_mean_squared_error):", best_score)


Best Hyperparameters: {'subsample': 0.9, 'n_estimators': 240, 'min_child_weight': 4.5, 'max_depth': 4, 'learning_rate': 0.1, 'gamma': 0.35, 'colsample_bytree': 1.0}
Best Score (neg_mean_squared_error): -181650940.0439008


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
305 fits failed out of a total of 500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
200 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.10/dist-packages/xgboost/core.py", line 730, in inner_f
    return func(**kwargs)
  File "/usr/local/lib/python3.10/dist-packages/xgboost/sklearn.py", line 1090, in fit
    self._Booster = train(
  File "/usr/local/lib/python3.10/dist-packages/xgboost/core.py", line 730, in inner_f

In [ ]:
# Getting the best model
best_gb_model = random_search.best_estimator_
print(f"Best XGBoost Regression Model: ")
print(best_gb_model)

# Making predictions with the best model
best_gb_preds = best_gb_model.predict(X_test)

Best XGBoost Regression Model: 
XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=1.0, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=0.35, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.1, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=4, max_leaves=None,
             min_child_weight=4.5, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=240, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)


In [ ]:
# Evaluating the best model
evaluate_model('XGBoost Regression (Tuned)', Y_test, best_gb_preds)

XGBoost Regression (Tuned) Model:
Mean Absolute Error: 9737.47
Mean Squared Error: 246797911.72
R-squared Score: 0.83
------------------------------------------------------------


In [ ]:
# Storing the model
model = XGBRegressor(subsample=0.95, n_estimators=240, min_child_weight=3.5, max_depth=4, learning_rate=0.2, gamma=0.35, colsample_bytree=1.0)
# model = GradientBoostingRegressor(learning_rate=0.2, max_depth=4, min_samples_split=10, n_estimators=200)
model.fit(X, Y)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=1.0, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=0.35, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.2, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=4, max_leaves=None,
             min_child_weight=3.5, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=240, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

# Creating Streamlit UI application for our laptop price prediction model

***
Creating a data frame of the features column only by excluding target column (price) from this new_df, so we it will be helpful for our streamlit application input elements
***
and also removing the  gpu model series column as it creating high dimensionality

In [ ]:
feature_df = new_df.drop(columns=['Price'])

In [ ]:
feature_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1272 entries, 0 to 1271
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1272 non-null   object 
 1   TypeName          1272 non-null   object 
 2   Inches            1272 non-null   float64
 3   Weight_kg         1272 non-null   float64
 4   IPS_Panel         1272 non-null   object 
 5   Touchscreen       1272 non-null   object 
 6   PPI               1272 non-null   int64  
 7   Display_Type      1272 non-null   object 
 8   CPU_Speed_GHz     1272 non-null   float64
 9   RAM_GB            1272 non-null   int64  
 10  Memory_1_Size_GB  1272 non-null   int64  
 11  Memory_1_Type     1272 non-null   object 
 12  Memory_2_Size_GB  1272 non-null   int64  
 13  Memory_2_Type     1272 non-null   object 
 14  GPU_Model_series  1272 non-null   object 
 15  OS_Brand          1272 non-null   object 
 16  OS_Version        1272 non-null   object 


In [ ]:
# Save the DataFrame to an Excel file
feature_df.to_excel('feature_df_laptop.xlsx', index=False)

In [ ]:
# !pip install streamlit

In [ ]:
# Serialize and save the trained model to a file
import joblib
joblib.dump(model, 'best_xgb_model.pkl')

['best_xgb_model.pkl']

##Streamlit app development of laptop price prediction
we have used below code in pycharm as .py file

In [ ]:
import streamlit as st
import pandas as pd
import joblib
from sklearn.preprocessing import LabelEncoder
import xgboost
import numpy as np

# Load the trained model
model = joblib.load('best_xgb_model.pkl')

# loading the dataset
feature_data = pd.read_excel("feature_df_laptop.xlsx")

# Function to preprocess input features
def preprocess_input(features, label_encoders):
    # Label encoding for categorical features
    for column, encoder in label_encoders.items():
        if column in ['IPS_Panel', 'Touchscreen']:
            # Convert 'Yes'/'No' to 1/0
            features[column] = 1 if features[column] == 'Yes' else 0
        else:
            features[column] = encoder.transform([features[column]])[0]
    return features

# Define label encoders for categorical features
label_encode_columns = ['Company', 'TypeName', 'IPS_Panel', 'Touchscreen', 'Memory_1_Type', 'Memory_2_Type', 'OS_Brand', 'OS_Version', 'Processor', 'Graphic_Card', 'GPU_Model_series', 'Display_Type']
label_encoders = {}
for column in label_encode_columns:
    label_encoders[column] = LabelEncoder()
    label_encoders[column].fit(feature_data[column])

# Create the Streamlit app
def main():
    st.title('Laptop Price Predictor')

    # Define input fields
    Company = st.selectbox('Company', feature_data['Company'].unique())
    TypeName = st.selectbox('Type Name', feature_data['TypeName'].unique())
    Inches = st.slider('Inches', min_value=int(feature_data['Inches'].min()), max_value=int(feature_data['Inches'].max()), value=int(feature_data['Inches'].mean()))
    Weight_kg = st.slider('Weight (kg)', min_value=float(feature_data['Weight_kg'].min()), max_value=float(feature_data['Weight_kg'].max()), value=float(feature_data['Weight_kg'].mean()))
    IPS_Panel = st.radio('IPS Panel', ['Yes', 'No'])
    Touchscreen = st.radio('Touchscreen', ['Yes', 'No'])
    PPI = st.slider('PPI', min_value=int(feature_data['PPI'].min()), max_value=int(feature_data['PPI'].max()), value=int(feature_data['PPI'].mean()))
    Display_Type = st.selectbox('Display Type', feature_data['Display_Type'].unique())
    CPU_Speed_GHz = st.slider('CPU Speed (GHz)', min_value=float(feature_data['CPU_Speed_GHz'].min()), max_value=float(feature_data['CPU_Speed_GHz'].max()), value=float(feature_data['CPU_Speed_GHz'].mean()))
    RAM_GB = st.slider('RAM (GB)', min_value=int(feature_data['RAM_GB'].min()), max_value=int(feature_data['RAM_GB'].max()), value=int(feature_data['RAM_GB'].mean()))
    Memory_1_Size_GB = st.slider('Memory 1 Size (GB)', min_value=int(feature_data['Memory_1_Size_GB'].min()), max_value=int(feature_data['Memory_1_Size_GB'].max()), value=int(feature_data['Memory_1_Size_GB'].mean()))
    Memory_1_Type = st.selectbox('Memory 1 Type', feature_data['Memory_1_Type'].unique())
    Memory_2_Size_GB = st.slider('Memory 2 Size (GB)', min_value=int(feature_data['Memory_2_Size_GB'].min()), max_value=int(feature_data['Memory_2_Size_GB'].max()), value=int(feature_data['Memory_2_Size_GB'].mean()))
    Memory_2_Type = st.selectbox('Memory 2 Type', feature_data['Memory_2_Type'].unique())
    OS_Brand = st.selectbox('OS Brand', feature_data['OS_Brand'].unique())
    OS_Version = st.selectbox('OS Version', feature_data['OS_Version'].unique())
    Processor = st.selectbox('Processor', feature_data['Processor'].unique())
    Graphic_Card = st.selectbox('Graphic Card', feature_data['Graphic_Card'].unique())
    GPU_Model_series = st.selectbox("Graphic card's  Series", feature_data['GPU_Model_series'].unique())

    # Prepare input features as a dictionary
    input_features = {
        'Company': Company,
        'TypeName': TypeName,
        'Inches': Inches,
        'Weight_kg': Weight_kg,
        'IPS_Panel': IPS_Panel,
        'Touchscreen': Touchscreen,
        'PPI': PPI,
        'Display_Type': Display_Type,
        'CPU_Speed_GHz': CPU_Speed_GHz,
        'RAM_GB': RAM_GB,
        'Memory_1_Size_GB': Memory_1_Size_GB,
        'Memory_1_Type': Memory_1_Type,
        'Memory_2_Size_GB': Memory_2_Size_GB,
        'Memory_2_Type': Memory_2_Type,
        'OS_Brand': OS_Brand,
        'OS_Version': OS_Version,
        'Processor': Processor,
        'Graphic_Card': Graphic_Card,
        'GPU_Model_series' : GPU_Model_series
    }

    # Preprocess input features
    input_features = preprocess_input(input_features, label_encoders)

    # Convert input_features to a numpy array
    input_features = np.array(list(input_features.values()))

    # Make predictions
    prediction = model.predict(input_features.reshape(1, -1))

    # Display prediction to user
    st.write(f"<h1 style='font-size: 30px;'>Predicted Laptop Price: INR {prediction[0]}</h1>", unsafe_allow_html=True)


if __name__ == '__main__':
    main()


In [ ]:
# to run this app on notebook
# %run laptop_price_predictor.py

In [ ]:
# to run app on command
# streamlit run laptop_price_predictor.py
